feat: add `lapack/base/dlarf1f` (#7540)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dlarf1f/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlarf1f

> 

Reviewer: You're missing the package description here.

Reviewer: Ah, I see what you did.

> > Apply a real elementary reflector `H = I - tau * v * v ^ T` to a real M by N matrix `C`.

Reviewer: ```suggestion
> Apply a real elementary reflector `H = I - tau * v * v^T` to a real M by N matrix `C`.
```

Applies here and throughout this PR. This follows conventions used elsewhere in stdlib regarding the transpose of a matrix/vector.

<section class="intro">

In linear algebra, a `Householder transformation` (or an `elementary reflector`) is a linear transformation that describes a reflection about a plane or a hyperplane containing the origin.

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var dlarf1f = require( '@stdlib/lapack/base/dlarf1f' );
```

#### dlarf1f( order, side, M, N, V, incv, tau, C, LDC, work )

Applies a real elementary reflector `H = I - tau * v * v ^ T` to a real M by N matrix `C`.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var C = new Float64Array( [ 1.0, 5.0, 9.0, 2.0, 6.0, 10.0, 3.0, 7.0, 11.0, 4.0, 8.0, 12.0 ] );
var V = new Float64Array( [ 0.5, 0.5, 0.5, 0.5 ] );
var work = new Float64Array( 3 );

var out = dlarf1f( 'row-major', 'left', 4, 3, V, 1, 1.0, C, 3, work );
// returns <Float64Array>[ -4.5, -10.5, -16.5, -0.75, -1.75, -2.75, 0.25, -0.75, -1.75, 1.25,  0.25, -0.75 ]
```

The function has the following parameters:

-   **order**: storage layout.

> -   **side**: specifies the side of multiplication with `C`. Use `left` to form `H * C` and `right` to from `C * H`.

Reviewer: ```suggestion
-   **side**: specifies the side of multiplication with `C`. Use `'left'` to form `H * C` and `'right'` to form `C * H`.
```

-   **M**: number of rows in `C`.
-   **N**: number of columns in `C`.

> -   **V**: the vector `v` in the representation of `H` as a [`Float64Array`][mdn-float64array].

Reviewer: ```suggestion
-   **V**: the vector `v` as a [`Float64Array`][mdn-float64array].
```

-   **incv**: stride length for `V`. If `incv` is negative, the elements of `V` are accessed in reverse order.

> -   **tau**: the value of `tau` in representation of `H`.

Reviewer: ```suggestion
-   **tau**: scalar constant.
```

-   **C**: input matrix stored in linear memory as a [`Float64Array`][mdn-float64array].
-   **LDC**: stride of the first dimension of `C` (a.k.a., leading dimension of the matrix `C`).
-   **work**: workspace [`Float64Array`][mdn-float64array].

Based on the `side` of multiplication, the expected storage of the array changes:

-   `work` should have `N` indexed elements if `side` = `left` and `M` indexed elements if `side` = `right`.

> -   `V` should have `1 + (M-1) * abs(incv)` indexed elements if `side` = `left` and `1 + (N-1) * abs(incv)` indexed elements if `side` = `right`.

Reviewer: ```suggestion
When `side` is `'left'`,

-   `work` should have `N` indexed elements.
-   `V` should have `1 + (M-1) * abs(incv)` indexed elements.

When `side` is `'right'`,

-   `work` should have `M` indexed elements.
-   `V` should have `1 + (N-1) * abs(incv)` indexed elements.
```

The sign of the increment parameter `incv` determines the order in which elements of `V` are accessed. For example, to access elements in reverse order,

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var C = new Float64Array( [ 1.0, 5.0, 9.0, 2.0, 6.0, 10.0, 3.0, 7.0, 11.0, 4.0, 8.0, 12.0 ] );
var V = new Float64Array( [ 0.5, 0.5, 0.5, 0.5 ] );
var work = new Float64Array( 3 );

var out = dlarf1f( 'row-major', 'left', 4, 3, V, -1, 1.0, C, 3, work );

> // returns <Float64Array>[ -4.5, -10.5, -16.5, -0.75, -1.75, -2.75, 0.25, -0.75, -1.75, 1.25,  0.25, -0.75 ]

Reviewer: This example is not super informative, as `V` is symmetric. Whether you go forward or reverse, you get the same values. I suggest updating this example to make it more meaningful.

var V1 = new Float64Array( V0.buffer, V0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var work1 = new Float64Array( work0.buffer, work0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

var our = dlarf1f( 'row-major', 'left', 4, 3, V1, 1, 1.0, C1, 3, work1 );
// C0 => <Float64Array>[ 0.0, -4.5, -10.5, -16.5, -0.75, -1.75, -2.75, 0.25, -0.75, -1.75, 1.25,  0.25, -0.75 ]
```

#### dlarf1f.ndarray( side, M, N, V, sv, ov, tau, C, sc1, sc2, oc, work, sw, ow )

Applies a real elementary reflector `H = I - tau * v * v ^ T` to a real M by N matrix `C` using alternative indexing semantics.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var C = new Float64Array( [ 1.0, 5.0, 9.0, 2.0, 6.0, 10.0, 3.0, 7.0, 11.0, 4.0, 8.0, 12.0 ] );
var V = new Float64Array( [ 0.5, 0.5, 0.5, 0.5 ] );
var work = new Float64Array( 3 );

var out = dlarf1f.ndarray( 'left', 4, 3, V, 1, 0, 1.0, C, 3, 1, 0, work, 1, 0 );
// returns <Float64Array>[ -4.5, -10.5, -16.5, -0.75, -1.75, -2.75, 0.25, -0.75, -1.75, 1.25,  0.25, -0.75 ]
```

The function has the following additional parameters:

> -   **side**: specifies the side of multiplication with `C`. Use `left` to form `H * C` and `right` to from `C * H`.

Reviewer: Same edits here as done above.

-   **M**: number of rows in `C`.
-   **N**: number of columns in `C`.
-   **V**: the vector `v` in the representation of `H` as a [`Float64Array`][mdn-float64array].
-   **sv**: stride length for `V`.
-   **ov**: starting index for `V`.
-   **tau**: the value of `tau` in representation of `H`.
-   **C**: input matrix as a [`Float64Array`][mdn-float64array].
-   **sc1**: stride of the first dimension of `C`.
-   **sc2**: stride of the second dimension of `C`.
-   **oc**: starting index for `C`.
-   **work**: workspace array as a [`Float64Array`][mdn-float64array].
-   **sw**: stride length for `work`.
-   **ow**: starting index for `work`.

Based on the `side` of multiplication, the expected storage of the array changes:

> -   `work` should have `N` indexed elements if `side` = `left` and `M` indexed elements if `side` = `right`.

Reviewer: Same comments as above.
